src/GitRepository.cc

```diff
@@ -0,0 +1,415 @@
+
+    // create commit
+    int error = git_commit_create(
+        &commit_id,
+        repo_.get(),
+        "HEAD",
+        my_signature_.get(),
+        my_signature_.get(),
+        "UTF-8",
+        commit_message.c_str(),
+        tree.get(),
+        1,
+        &parent_commit
+    );
+
+    if (error) throw task::Error("Cannot commit.");
+}
+
+
+void GitRepository::add()
+{
+    //load index of last commit
+    LibGitPointer<git_index> gindex{repository_index(repo_.get())};
+
+    char *paths[] = {const_cast<char*>("*")};
+    git_strarray array = {paths, 1};
+
+    //add all
+    int error = git_index_add_all(gindex.get(), &array, GIT_INDEX_ADD_DEFAULT, nullptr, nullptr);
+    if (error) throw task::Error("Cannot stage files.");
+
+    // save addition
+    git_index_write(gindex.get());
+}
+
+
+void GitRepository::remove_directory(std::filesystem::path seq_directory)
+{
+    //load index of last commit
+    LibGitPointer<git_index> gindex{repository_index(repo_.get())};
+
+    //remove files from directory
+    int error = git_index_remove_directory(gindex.get(), seq_directory.c_str(), 0);
+    if (error) throw task::Error("Cannot remove sequence files.");
+    // remove sequence directory from git
+    error = git_index_remove_bypath(gindex.get(), (std::string(seq_directory)+"/").c_str());
+    if (error) throw task::Error("Cannot remove sequence directory.");
+
+    git_index_write(gindex.get());
+    //git_index_write_tree(&tree, gindex.get());
+
+    // delete sequence
+    std::filesystem::remove_all(repo_path_ / seq_directory);
+}
+
+
+git_commit* GitRepository::get_commit(int count)
+{
+    std::string ref = "HEAD^" + std::to_string(count);
+    return get_commit(ref);
+}
+git_commit* GitRepository::get_commit(const std::string& ref)
+{
+    git_commit * commit;
+    git_oid oid_parent_commit;
+
+    // resolve HEAD into a SHA1
+    int error = git_reference_name_to_id( &oid_parent_commit, repo_.get(), ref.c_str());
+    if (error) throw task::Error("Cannot find HEAD of branch.");
+
+    // find commit object by commit ID
+    error = git_commit_lookup( &commit, repo_.get(), &oid_parent_commit );
+    if (error) throw task::Error("Cannot find HEAD of branch.");
+
+    return commit;
+}
+git_commit* GitRepository::get_commit()
+{
+    return get_commit(std::string{"HEAD"});
+}
+
+
+std::vector<FileStatus> GitRepository::collect_status(git_status_list *status) const
+{
+    // get number of submodules
+    const size_t nr_entries = git_status_list_entrycount(status);
+
+    // declare status holding vector for each submodule
+    std::vector<FileStatus> return_array;
+    FileStatus filestats;
+
+    for (size_t i = 0; i < nr_entries; ++i)
+    {
+        const git_status_entry *s = nullptr;
+        const char *old_path = nullptr;
+        const char *new_path = nullptr;
+
+        std::string istatus = "";
+        std::string wstatus = "";
+
+        s = git_status_byindex(status, i);
```

Comment:
Maybe the entire `s`-into-`FileStatus` conversion could be refactored into a function.

Reply:
is not it like this? (collect_status)